### PR TITLE
Add ONNX export support for TE modules

### PR DESCRIPTION
**Pull request by asfiyab-nvidia**
* Add TorchScript Operators
* Add symbolic methods to ONNX exporter
* Add tests for the ONNX export

Signed-off-by: Asfiya Baig <asfiyab@nvidia.com>
Signed-off-by: Neta Zmora <nzmora@nvidia.com>

### CODE REVIEW

**Comment by ptrendx** [comment]
Hi @asfiyab-nvidia, what is that libcustom so file?

**Comment by ptrendx** [comment]
Does that have to be closed source? If so, can we at least move it to tests directory instead of the top level one? If it does not have to be closed source then maybe we can have the source inside tests and compile it on the fly?

File: transformer_engine/pytorch/csrc/ts_fp8_op.cpp [hunk 1]
```diff
@@ -0,0 +1,179 @@
+/*************************************************************************
+ * Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ *
+ * See LICENSE for license information.
+ ************************************************************************/
+
+#include <torch/script.h>
+#include "extensions.h"
+
+transformer_engine::DType reverse_map_dtype(int64_t dtype) {
```
Comment (ptrendx) [inline transformer_engine/pytorch/csrc/ts_fp8_op.cpp:10]: Hmm, can we make it just a static_cast + boundary condition ensuring that it is valid? While I don't expect adding new types very often, this switch is a potential source of problems if we do add a new type.

Comment (ptrendx) [reply transformer_engine/pytorch/csrc/ts_fp8_op.cpp:10]: Also, could you put this function in an unnamed namespace maybe?

File: transformer_engine/pytorch/module.py [hunk 2]
```diff
@@ -162,10 +166,12 @@ def get_extra_state(self) -> Union[List[Any], None]:
                     extra[k] = v
             state["extra_fp8_variables"] = extra
 
-            return state
-        return None
+        state_serialized = pickle.dumps(state)
+        state_tensor = torch.tensor(np.frombuffer(state_serialized, dtype=np.uint8), device='cuda')
```
Comment (ptrendx) [inline transformer_engine/pytorch/module.py:170]: Why do you put this tensor to GPU?

File: transformer_engine/pytorch/csrc/ts_fp8_op.cpp (hunk 1 above)
Comment (asfiyab-nvidia) [reply transformer_engine/pytorch/csrc/ts_fp8_op.cpp:10]: I have taken care of the remaining comments locally and will push the changes shortly. 

> can we make it just a static_cast + boundary condition ensuring that it is valid

Quick question on this comment - are you suggesting that the switch statement to be removed? Can you elaborate on how we can achieve the int to dtype mapping in that case?

Comment (ptrendx) [reply transformer_engine/pytorch/csrc/ts_fp8_op.cpp:10]: Yeah, I thought about something like
```
if (dtype >= 0 && dtype < largest_value) {
  return static_cast<transformer_engine::DType>(dtype);
} else {
  ERROR
}
```
We would probably need to add then some `DType::Num_Types` type to the enum though.

Comment (asfiyab-nvidia) [reply transformer_engine/pytorch/csrc/ts_fp8_op.cpp:10]: Ah I see what you mean - will make this change. I do see a `kNumTypes` in the enum already. Let me test it and push the change.

Comment (asfiyab-nvidia) [reply transformer_engine/pytorch/csrc/ts_fp8_op.cpp]: Addressed both comments in the latest commits

File: transformer_engine/pytorch/module.py (hunk 2 above)
Comment (asfiyab-nvidia) [reply transformer_engine/pytorch/module.py]: Addressed

**Comment by netaz** [comment]
@ptrendx is there some code in TE we can leverage to query the SM version, or do you recommend us installing some lib (e.g. pynvml)?

**Comment by asfiyab-nvidia** [comment]
@ptrendx can you please authorize a pipeline run for the latest commit? It contains fixes for the failures from the last run. Thanks

File: transformer_engine/pytorch/__init__.py [hunk 3]
```diff
@@ -10,3 +10,5 @@
 from .transformer import TransformerLayer
 from .fp8 import fp8_autocast
 from .distributed import checkpoint
+# Register custom op symbolic ONNX functions
+from .te_onnx_extensions import *
```
Comment (ksivaman) [inline transformer_engine/pytorch/__init__.py:14]:
```suggestion
from .te_onnx_extensions import (
    onnx_cast_to_fp8,
    onnx_cast_from_fp8,
    onnx_fp8_gelu,
    onnx_te_gemm,
    onnx_layernorm_fwd_fp8,
    onnx_layernorm_fwd,
)
```

Comment (ksivaman) [reply transformer_engine/pytorch/__init__.py:14]: Importing only what's needed

File: tests/test_onnx_export.py [hunk 4]
```diff
@@ -0,0 +1,955 @@
...
+import numpy as np
+import math
+import onnxruntime as ort
+import torch
+from torch import nn as nn
+from typing import Union, Tuple
+import transformer_engine.pytorch as te
+from transformer_engine.common import recipe
+import transformer_engine_extensions as tex
+from transformer_engine.pytorch.cpp_extensions import *
```
[the remark this answers is not in the document]
Comment (ksivaman) [reply tests/test_onnx_export.py:21]: Importing only what's needed

File: tests/test_onnx_export.py [hunk 5]
```diff
@@ -0,0 +1,955 @@
...
+            return_layernorm_output=return_layernorm_output,
+            params_dtype=precision,
+        ).to(device='cuda')
+        if use_fp8:
+            set_layer_scale(model, scale_factor)
+        do_export(model, inp, fname, use_fp8)
+        if not use_fp8:
+            validate_result(fname, inp, model, atol=1e-3)
+        elif precision not in (torch.bfloat16,):
+            validate_result(fname, inp, model, atol=1e-3, is_fp8=use_fp8)
```
Comment (ksivaman) [inline tests/test_onnx_export.py:686]:
```suggestion
            validate_result(fname, inp, model, atol=1e-2, is_fp8=use_fp8)
```

Comment (ksivaman) [reply tests/test_onnx_export.py:686]: With the previous tolerance, this test locally fails >50% of time.

**Review by ksivaman — COMMENTED** [review]
Initial comments

**Review by ksivaman — APPROVED** [review]
LGTM

**Pull request merged**